core/src/multipart_form_data.cpp

```diff
@@ -0,0 +1,179 @@
+#include "irods/private/http_api/multipart_form_data.hpp"
+
+#include "irods/private/http_api/log.hpp"
+
+#include <boost/algorithm/string.hpp>
+#include <boost/beast/http/rfc7230.hpp>
+
+#include <fmt/format.h>
+
+namespace irods::http
+{
+	auto get_multipart_form_data_boundary(const std::string_view _data) -> std::optional<std::string_view>
+	{
+		// The ext_list type does not treat "multipart/form-data" as a single string.
+		// I'm not sure if this is by design, but we have to skip the multipart/ prefix so that
+		// ext_list can extract the boundary argument.
+		//
+		// TODO Consider opening an issue for Boost.Beast in Github.
+
+		const auto pos = _data.find("form-data");
+
+		if (pos == std::string_view::npos) {
+			return std::nullopt;
+		}
+
+		boost::beast::http::ext_list list{_data.substr(pos)};
+		const auto iter = list.find("form-data");
+
+		if (iter == std::end(list)) {
+			log::error("{}: Missing [boundary] parameter in [multipart/form-data] value.", __func__);
+			return std::nullopt;
+		}
+
+		for (auto&& param : iter->second) {
+			if (param.first == "boundary") {
+				log::debug("PARAM = (name={}, value={})", param.first, param.second);
+				return param.second;
+			}
+		}
+
+		return std::nullopt;
+	} // get_multipart_form_data_boundary
+
+	auto parse_multipart_form_data(const std::string_view _boundary, const std::string_view _data)
+		-> query_arguments_type
+	{
+		if (_data.empty()) {
+			return {};
+		}
+
+		const auto boundary_start = fmt::format("--{}", _boundary);
+		const auto boundary_end = fmt::format("{}--", boundary_start);
+
+		enum class parser_state
+		{
+			find_boundary_start,
+			read_boundary_start,
+			read_header,
+			read_body
```

Comment:
Consider adding a few words describing each parser state.

Reply:
This isn't necessary. The enum names describe their purpose enough for people to follow along.